File: src/main/java/com/lessmarkup/interfaces/module/ModuleProvider.java
package com.lessmarkup.interfaces.module;

import com.lessmarkup.interfaces.data.DomainModelProvider;
import com.lessmarkup.interfaces.structure.NodeHandler;
import com.lessmarkup.interfaces.structure.Tuple;

import java.util.Collection;

public interface ModuleProvider {
    Collection<ModuleConfiguration> getModules();
    void discoverAndRegisterModules();
    void updateModuleDatabase(DomainModelProvider domainModelProvider);
    Collection<String> getNodeHandlers();
    Tuple<Class<? extends NodeHandler>, String> getNodeHandler(String id);
}

File: src/main/scala/com/lessmarkup/interfaces/module/ModuleProvider.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public License, v. 2.0.
 * If a copy of the MPL was not distributed with this file, You can obtain one at
 * http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.interfaces.module

import com.lessmarkup.interfaces.data.DomainModelProvider
import com.lessmarkup.interfaces.structure.NodeHandlerFactory

trait ModuleProvider {
  def getModules: Seq[ModuleConfiguration]
  def updateModuleDatabase(domainModelProvider: DomainModelProvider)
  def getNodeHandlers: Seq[String]
  def getNodeHandler(id: String): Option[(Class[_ <: NodeHandlerFactory], String)]
}
